pkg/kubefnord/unjoin.go

```diff
@@ -0,0 +1,387 @@
+		# cluster name and the context name of the federation
+		# control plane's host cluster. Cluster name must be
+		# a valid RFC 1123 subdomain name. Cluster context
+		# must be specified if the cluster name is different
+		# than the cluster's context in the local kubeconfig.
+		kubefnord unjoin foo --host-cluster-context=bar`
+)
+
+type unjoinFederation struct {
+	options.SubcommandOptions
+	unjoinFederationOptions
+}
+
+type unjoinFederationOptions struct {
+	clusterContext     string
+	removeFromRegistry bool
+}
+
+// Bind adds the unjoin specific arguments to the flagset passed in as an
+// argument.
+func (o *unjoinFederationOptions) Bind(flags *pflag.FlagSet) {
+	flags.StringVar(&o.clusterContext, "cluster-context", "",
+		"Name of the cluster's context in the local kubeconfig. Defaults to cluster name if unspecified.")
+	flags.BoolVar(&o.removeFromRegistry, "remove-from-registry", false,
+		"Remove the cluster from the cluster registry that is aggregated with the kubernetes API server running in the host cluster context.")
+}
+
+// NewCmdUnJoin defines the `unjoin` command that unjoins a cluster from a
+// federation.
+func NewCmdUnJoin(cmdOut io.Writer, config util.FedConfig) *cobra.Command {
+	opts := &unjoinFederation{}
+
+	cmd := &cobra.Command{
+		Use:     "unjoin CLUSTER_NAME --host-cluster-context=HOST_CONTEXT",
+		Short:   "Unjoin a cluster from a federation",
+		Long:    unjoin_long,
+		Example: unjoin_example,
+		Run: func(cmd *cobra.Command, args []string) {
+			err := opts.Complete(args)
+			if err != nil {
+				glog.Fatalf("error: %v", err)
+			}
+
+			err = opts.Run(cmdOut, config)
+			if err != nil {
+				glog.Fatalf("error: %v", err)
+			}
+		},
+	}
+
+	flags := cmd.Flags()
+	opts.CommonBind(flags)
+	opts.Bind(flags)
+
+	return cmd
+}
+
+// Complete ensures that options are valid and marshals them if necessary.
+func (j *unjoinFederation) Complete(args []string) error {
+	err := j.SetName(args)
+	if err != nil {
+		return err
+	}
+
+	if j.clusterContext == "" {
+		glog.V(2).Infof("Defaulting cluster context to unjoining cluster name %s", j.ClusterName)
+		j.clusterContext = j.ClusterName
+	}
+
+	glog.V(2).Infof("Args and flags: name %s, host: %s, host-system-namespace: %s, kubeconfig: %s, cluster-context: %s, dry-run: %v",
+		j.ClusterName, j.Host, j.FederationNamespace, j.Kubeconfig, j.clusterContext, j.DryRun)
+
+	return nil
+}
+
+// Run is the implementation of the `unjoin federation` command.
+func (j *unjoinFederation) Run(cmdOut io.Writer, config util.FedConfig) error {
+	hostConfig, err := config.HostConfig(j.Host, j.Kubeconfig)
+	if err != nil {
+		// TODO(font): Return new error with this same text so it can be output
+		// by caller.
+		glog.V(2).Infof("Failed to get host cluster config: %v", err)
+		return err
+	}
+
+	clusterConfig, err := config.ClusterConfig(j.clusterContext, j.Kubeconfig)
+	if err != nil {
+		glog.V(2).Infof("Failed to get joining cluster config: %v", err)
```

Comment:
s/joining/unjoining
I'm not sure that using "unjoining" in this way is ideal. It is consistent with the join command though! 
@marun, thoughts?

Reply:
sgtm